From ef7dbac4473ad98ec8d06463b9fad57b138fc331 Mon Sep 17 00:00:00 2001
From: Michael Campion <michael.campion@emutex.com>
Date: Wed, 20 Nov 2019 13:15:08 +0000
Subject: [PATCH 2/3] platform/x86: upboard: Add gpio chip select for UP Board
 SPI

The UP Board SOC provides only 1 native chip select but requires 2 chip
select lines for compatibility with RPi peripherals.

This patch adds the option to enable a gpio chip select on any available
HAT gpio (0 to 27) by passing the value to the up_board module parameter
spi_cs1_gpio on the kernel command line.

Signed-off-by: Michael Campion <michael.campion@emutex.com>
---
 drivers/platform/x86/up_board.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/drivers/platform/x86/up_board.c b/drivers/platform/x86/up_board.c
index 3acbed7..a83ca94 100644
--- a/drivers/platform/x86/up_board.c
+++ b/drivers/platform/x86/up_board.c
@@ -28,6 +28,8 @@
 #include <linux/regulator/fixed.h>
 #include <linux/regulator/machine.h>
 #include <linux/acpi.h>
+#include <linux/gpio/machine.h>
+#include <linux/gpio.h>
 
 #define UP_BOARD_SPIDEV_BUS_NUM 2
 #define UP_BOARD_SPIDEV_MAX_CLK 25000000
@@ -46,6 +48,14 @@ static bool spidev1 = false;
 module_param(spidev1, bool, S_IRUGO);
 MODULE_PARM_DESC(spidev1, "register a spidev device on SPI bus 2-1");
 
+/* Default SPI bus 2-1 chip select on HAT pin 26 */
+#define SPI_CS1_DEFAULT_GPIO (7)
+
+static int spi_cs1_gpio = SPI_CS1_DEFAULT_GPIO;
+module_param(spi_cs1_gpio, int, S_IRUGO);
+MODULE_PARM_DESC(spi_cs1_gpio,
+		"Gpio for SPI bus 2-1 chip select (0 to 27), or -1 to disable");
+
 /* On the UP board, if the ODEn bit is set on the pad configuration
  * it seems to impair some functions on the I/O header such as UART, SPI
  * and even I2C.  So we disable it for all header pins by default.
@@ -101,6 +111,19 @@ static struct spi_board_info up_spidev_info __initdata = {
 	.max_speed_hz   = UP_BOARD_SPIDEV_MAX_CLK,
 };
 
+/* On UP Board there is no native chip select 1 available on SPI Bus 2. This
+ * table will be used to reserve a gpio for chip select 1 using the default
+ * gpio on HAT Pin 26 or the gpio number passed to this module's spi_cs1_gpio
+ * parameter. -1 will disable the reservation of a gpio for chip select.
+ */
+static struct gpiod_lookup_table up_spi_cs_gpiod_table = {
+	.dev_id		= "8086228E:01",
+	.table		= {
+		GPIO_LOOKUP_IDX("up-pinctrl",
+				SPI_CS1_DEFAULT_GPIO, "cs", 1, GPIO_ACTIVE_HIGH),
+	},
+};
+
 static int __init
 up_board_init_devices(void)
 {
@@ -170,6 +193,11 @@ up_board_init(void) {
 
 	board_info = system_id->driver_data;
 
+	if (gpio_is_valid(spi_cs1_gpio)) {
+		up_spi_cs_gpiod_table.table[0].chip_hwnum = spi_cs1_gpio;
+		gpiod_add_lookup_table(&up_spi_cs_gpiod_table);
+	}
+
 	/* Register pin control mappings specific to board version */
 	if (board_info->pinmux_maps) {
 		ret = pinctrl_register_mappings(board_info->pinmux_maps,
-- 
2.7.4

